include/boost/math/statistics/univariate_statistics.hpp

```diff
@@ -12,75 +12,101 @@
 #include <cmath>
 #include <vector>
 #include <boost/assert.hpp>
+#include <atomic>
+#include <type_traits>
+#include <execution>
+#include <numeric>
+#include <valarray>
 
 namespace boost::math::statistics {
 
-template<class ForwardIterator>
-auto mean(ForwardIterator first, ForwardIterator last)
+template<class ExecutionPolicy, class ForwardIterator>
+auto mean(ExecutionPolicy&& exec, ForwardIterator first, ForwardIterator last)
 {
     using Real = typename std::iterator_traits<ForwardIterator>::value_type;
     BOOST_ASSERT_MSG(first != last, "At least one sample is required to compute the mean.");
-    if constexpr (std::is_integral<Real>::value)
+    if constexpr (std::is_integral_v<Real>)
     {
-        double mu = 0;
-        double i = 1;
-        for(auto it = first; it != last; ++it) {
-            mu = mu + (*it - mu)/i;
-            i += 1;
+        if constexpr (std::is_same_v<std::remove_reference_t<decltype(exec)>, decltype(std::execution::seq)>)
+        {
+            double mu = 0;
+            double i = 1;
+            for(auto it = first; it != last; ++it) {
+                mu = mu + (*it - mu)/i;
+                i += 1;
+            }
+            return mu;
+        }
+        else
+        {
+            return std::reduce(exec, first, last, 0.0) / std::distance(first, last);
         }
-        return mu;
     }
     else if constexpr (std::is_same_v<typename std::iterator_traits<ForwardIterator>::iterator_category, std::random_access_iterator_tag>)
     {
-        size_t elements = std::distance(first, last);
-        Real mu0 = 0;
-        Real mu1 = 0;
-        Real mu2 = 0;
-        Real mu3 = 0;
+        std::size_t elements {std::distance(first, last)};
+        std::valarray<Real> mu {0, 0, 0, 0};
+        std::valarray<Real> temp {0, 0, 0, 0};
```

Comment:
Isn't `valarray` on the verge of deprecation?

Reply:
Not that I am aware of. The implementation in the Intel MKL has some pretty serious optimization. It certainly has its idiosyncrasies but it is good at what it does.

Reply:
Cool; but let's check the generated asm on godbolt and make sure it looks good.